File: src/crew/resolveBackgroundCrafting.ts
import type { CrewMember } from "../data/types/crewMember.ts";
import type { CraftAction } from "../data/types/craftAction.ts";
import type { BackgroundResult } from "../data/types/backgroundResult.ts";
import type { RandomFn } from "../data/types/random.ts";
import { craft } from "../simulation/craft.ts";
import { ELAPSED_TIME_CAP_HOURS, BACKGROUND_IDLE_OUTPUT_RATE } from "../data/constants/crewConfig.ts";

const MS_PER_HOUR = 60 * 60 * 1000;

// Phase 4 GDD §2.1a. Elapsed time is always derived from
// currentTime - lastCheckedAt -- never trusts a caller-supplied duration,
// which is the entire point of storing lastCheckedAt rather than letting
// a client assert "N hours passed" (see crewMember.ts's own comment).
// `backgroundRate` is injectable (defaults to crewConfig.ts's
// BACKGROUND_IDLE_OUTPUT_RATE, now resolved to a real 0.5 units/hour --
// see that constant's own comment for the "flat 50%" design decision and
// why 0.5/hour is the concrete number it resolves to) rather than
// hardcoded, mirroring this codebase's existing injectable-with-a-default
// pattern (RandomFn, `now`). The explicit `null` override still exists
// and still produces the "not yet available" result below -- kept for a
// caller that deliberately wants to represent "no background production
// mechanism at all" (e.g. a future crew type this doesn't apply to),
// not because the rate itself is unresolved anymore.
export function resolveBackgroundCrafting(
  crewMember: CrewMember,
  craftAction: CraftAction,
  currentTime: number,
  backgroundRate: number | null = BACKGROUND_IDLE_OUTPUT_RATE,
  random?: RandomFn,
  // Real-inventory availability cap, pre-resolved by the caller -- same
  // "core function never touches Inventory directly, caller passes in
  // what's available" boundary buildCitadel()'s materialQuantityAvailable
  // already established. Defaults to unbounded so every pre-existing call
  // site (and every existing test) is unaffected. CrewScene computes this
  // from real inventory via totalQuantity() before calling, so a crew
  // member idle for long enough to time-compute more units than the
  // player's actual stockpile supports is correctly clamped down, never
  // producing output from materials that don't exist.
  maxUnits: number = Infinity,
): BackgroundResult {
  const updatedCrewMember: CrewMember = { ...crewMember, lastCheckedAt: currentTime };

  if (backgroundRate === null) {
    return {
      resolved: false,
      reason: "no background/idle output rate configured for this call",
      updatedCrewMember,
    };
  }

  const rawElapsedHours = (currentTime - crewMember.lastCheckedAt) / MS_PER_HOUR;
  const cappedElapsedHours = Math.min(Math.max(rawElapsedHours, 0), ELAPSED_TIME_CAP_HOURS);
  const unitsCompleted = Math.min(Math.floor(cappedElapsedHours * backgroundRate), maxUnits);

  const results = [];
  for (let i = 0; i < unitsCompleted; i++) {
    results.push(craft(craftAction.inputs, craftAction.recipe, craftAction.schematicTier, crewMember.tier, random));
  }

  return { resolved: true, unitsCompleted, results, updatedCrewMember };
}

File: unity/ProfitableCore/Simulation/ResolveBackgroundCrafting.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/crew/resolveBackgroundCrafting.ts. Elapsed time is always
// derived from currentTime - LastCheckedAt -- never trusts a caller-
// supplied duration, which is the entire point of storing LastCheckedAt.
public static class ResolveBackgroundCraftingSimulation
{
    private const double MsPerHour = 60 * 60 * 1000;

    // The TypeScript source's `backgroundRate: number | null =
    // BACKGROUND_IDLE_OUTPUT_RATE` default is evaluated PER CALL (a live
    // JS default-argument read), so an omitted argument always uses
    // whatever the config currently holds -- including a future value if
    // it changes at runtime. C# default parameter values are baked in at
    // compile time, so they can't read a mutable static property live;
    // this overload is the translation of that live-default behavior
    // instead. Passing an explicit `null` (not omitting the argument)
    // still means "no background output configured for this call" --
    // callers that want that must call the other overload directly.
    public static BackgroundResult ResolveBackgroundCrafting(
        CrewMember crewMember,
        CraftAction craftAction,
        long currentTimeMs,
        RandomFn? random = null,
        double maxUnits = double.PositiveInfinity)
        => ResolveBackgroundCrafting(crewMember, craftAction, currentTimeMs, CrewConfig.BackgroundIdleOutputRate, random, maxUnits);

    // `maxUnits` is the real-inventory availability cap, pre-resolved by
    // the caller -- same "core function never touches Inventory directly,
    // caller passes in what's available" boundary this project's own
    // BuildCitadel-equivalent reasoning already established. Defaults to
    // unbounded so a caller with no inventory constraint to apply is
    // unaffected.
    public static BackgroundResult ResolveBackgroundCrafting(
        CrewMember crewMember,
        CraftAction craftAction,
        long currentTimeMs,
        double? backgroundRate,
        RandomFn? random,
        double maxUnits)
    {
        var updatedCrewMember = new CrewMember
        {
            Id = crewMember.Id,
            HiredByPlayerId = crewMember.HiredByPlayerId,
            Tier = crewMember.Tier,
            Profession = crewMember.Profession,
            Status = crewMember.Status,
            AssignedCraftId = crewMember.AssignedCraftId,
            HiredAt = crewMember.HiredAt,
            LastCheckedAt = currentTimeMs,
            WageAmount = crewMember.WageAmount,
            LastPaidAt = crewMember.LastPaidAt,
            UnavailableUntil = crewMember.UnavailableUntil,
            ShipRole = crewMember.ShipRole,
            AssignedShipId = crewMember.AssignedShipId,
        };

        if (backgroundRate is null)
        {
            return new BackgroundRateUnavailable
            {
                Reason = "no background/idle output rate configured for this call",
                UpdatedCrewMember = updatedCrewMember,
            };
        }

        var rawElapsedHours = (currentTimeMs - crewMember.LastCheckedAt) / MsPerHour;
        var cappedElapsedHours = Math.Min(Math.Max(rawElapsedHours, 0), CrewConfig.ElapsedTimeCapHours);
        var unitsCompleted = (int)Math.Min(Math.Floor(cappedElapsedHours * backgroundRate.Value), maxUnits);

        var results = new List<CraftResult>();
        for (var i = 0; i < unitsCompleted; i++)
        {
            results.Add(Crafter.Craft(craftAction.Inputs, craftAction.Recipe, craftAction.SchematicTier, crewMember.Tier, random));
        }

        return new BackgroundResolved { UnitsCompleted = unitsCompleted, Results = results, UpdatedCrewMember = updatedCrewMember };
    }
}
